File: packages/types/src/document.ts
import { z } from 'zod'

export const FileTypeSchema = z.enum([
  'PDF',
  'DOC',
  'DOCX',
  'TXT',
  'MD',
  'XLSX',
  'XLS',
  'CSV',
  'JPG',
  'JPEG',
  'PNG',
  'GIF',
  'WEBP',
  'MP4',
  'MOV',
  'AVI',
  'MKV',
])
export type FileType = z.infer<typeof FileTypeSchema>

export const ProcessingStatusSchema = z.enum(['PENDING', 'PROCESSING', 'READY', 'FAILED'])
export type ProcessingStatus = z.infer<typeof ProcessingStatusSchema>

export const DocumentSchema = z.object({
  id: z.string().cuid(),
  knowledgeBaseId: z.string().cuid(),
  filename: z.string().min(1),
  fileType: FileTypeSchema,
  fileSizeBytes: z.number().int().positive(),
  status: ProcessingStatusSchema,
  pageCount: z.number().int().positive().nullable(),
  errorMessage: z.string().nullable(),
  createdAt: z.coerce.date(),
})
export type Document = z.infer<typeof DocumentSchema>

export const DocumentStatusSchema = z.object({
  id: z.string().cuid(),
  status: ProcessingStatusSchema,
  pageCount: z.number().int().positive().nullable(),
  errorMessage: z.string().nullable(),
  progress: z
    .object({
      currentPage: z.number().int().nonnegative().nullable(),
      totalPages: z.number().int().positive().nullable(),
      stage: z.enum(['extracting', 'ocr', 'chunking', 'embedding']).nullable(),
    })
    .nullable(),
})
export type DocumentStatus = z.infer<typeof DocumentStatusSchema>

// Maps MIME types to FileType enum values — single source of truth for upload validation
export const SUPPORTED_MIME_TYPES = {
  'application/pdf': 'PDF',
  'application/msword': 'DOC',
  'application/vnd.openxmlformats-officedocument.wordprocessingml.document': 'DOCX',
  'text/plain': 'TXT',
  'text/markdown': 'MD',
  'application/vnd.ms-excel': 'XLS',
  'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet': 'XLSX',
  'text/csv': 'CSV',
  'image/jpeg': 'JPG',
  'image/png': 'PNG',
  'image/gif': 'GIF',
  'image/webp': 'WEBP',
  'video/mp4': 'MP4',
  'video/quicktime': 'MOV',
  'video/x-msvideo': 'AVI',
  'video/x-matroska': 'MKV',
} as const satisfies Record<string, FileType>

export type SupportedMimeType = keyof typeof SUPPORTED_MIME_TYPES

export const MAX_FILE_SIZE_BYTES = 500 * 1024 * 1024 // 500 MB

File: apps/api/app/models/document.py
import enum
from datetime import datetime
from typing import TYPE_CHECKING

from sqlalchemy import BigInteger, ForeignKey, Integer, String, Text, func
from sqlalchemy import Enum as SAEnum
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base

if TYPE_CHECKING:
    from app.models.knowledge_base import KnowledgeBaseModel


class FileType(str, enum.Enum):
    PDF = "PDF"
    DOC = "DOC"
    DOCX = "DOCX"
    TXT = "TXT"
    MD = "MD"
    XLSX = "XLSX"
    XLS = "XLS"
    CSV = "CSV"
    JPG = "JPG"
    JPEG = "JPEG"
    PNG = "PNG"
    GIF = "GIF"
    WEBP = "WEBP"
    MP4 = "MP4"
    MOV = "MOV"
    AVI = "AVI"
    MKV = "MKV"


class ProcessingStatus(str, enum.Enum):
    PENDING = "PENDING"
    PROCESSING = "PROCESSING"
    READY = "READY"
    FAILED = "FAILED"


class DocumentModel(Base):
    """
    Mirrors the Prisma `documents` table.
    Prisma owns the schema; SQLAlchemy reads rows and updates status during ingestion.
    """

    __tablename__ = "documents"

    id: Mapped[str] = mapped_column(String, primary_key=True)
    knowledge_base_id: Mapped[str] = mapped_column(
        String,
        ForeignKey("knowledge_bases.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    filename: Mapped[str] = mapped_column(String, nullable=False)
    # create_type=False — Prisma already created the enum in PostgreSQL
    file_type: Mapped[FileType] = mapped_column(
        SAEnum(FileType, name="FileType", create_type=False), nullable=False
    )
    file_size_bytes: Mapped[int] = mapped_column(BigInteger, nullable=False)
    status: Mapped[ProcessingStatus] = mapped_column(
        SAEnum(ProcessingStatus, name="ProcessingStatus", create_type=False),
        nullable=False,
        default=ProcessingStatus.PENDING,
    )
    page_count: Mapped[int | None] = mapped_column(Integer, nullable=True)
    error_message: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(server_default=func.now())

    knowledge_base: Mapped["KnowledgeBaseModel"] = relationship(
        "KnowledgeBaseModel",
        back_populates="documents",
    )

    def __repr__(self) -> str:
        return f"DocumentModel(id={self.id!r}, filename={self.filename!r}, status={self.status!r})"
